File: app/src/main/java/com/lovejjfg/arsenal/utils/FragmentsUtil.java
/*
 *  Copyright (c) 2017.  Joe
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package com.lovejjfg.arsenal.utils;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;

import com.lovejjfg.arsenal.base.SupportFragment;

import java.util.ArrayList;
import java.util.List;


/**
 * Created by Joe on 2016/11/13.
 * Email lovejjfg@gmail.com
 */

public class FragmentsUtil {
    private final FragmentManager manager;

    public FragmentsUtil(FragmentManager manager) {
        this.manager = manager;
    }

    public void addToShow(SupportFragment from, SupportFragment to) {
        bindContainerId(from.getContainerId(), to);
        FragmentTransaction transaction = manager.beginTransaction();
        String tag = to.getClass().getSimpleName();
        transaction.add(from.getContainerId(), to, tag)
                .addToBackStack(tag)
                .hide(from)
                .show(to)
                .commit();

    }

    public void addToParent(int containerViewId, @NonNull SupportFragment parent, int pos, SupportFragment... childs) {
        FragmentTransaction transaction = parent.getChildFragmentManager().beginTransaction();
        if (childs != null && childs.length > 0) {
            addFragmentsToStack(parent, containerViewId, pos, transaction, false, childs);
        }
    }

    public void replaceToParent(int containerViewId, @NonNull SupportFragment parent, SupportFragment... childs) {
        FragmentTransaction transaction = parent.getChildFragmentManager().beginTransaction();
        if (childs != null && childs.length > 0) {
            for (SupportFragment child : childs) {
                bindContainerId(containerViewId, child);
                String tag = child.getClass().getSimpleName();
                transaction.replace(containerViewId, child, tag)
                        .addToBackStack(tag);
            }
            transaction.commit();
        }
    }

    public void replaceToShow(SupportFragment from, SupportFragment to) {
        bindContainerId(from.getContainerId(), to);
        FragmentTransaction transaction = manager.beginTransaction();
        String tag = to.getClass().getSimpleName();
        transaction.replace(from.getContainerId(), to, tag)
                .addToBackStack(tag)
                .commit();

    }

    public void loadRoot(int containerViewId, int pos, SupportFragment... roots) {
        FragmentTransaction transaction = manager.beginTransaction();
        addFragmentsToStack(null, containerViewId, pos, transaction, true, roots);
    }

    private void addFragmentsToStack(SupportFragment parent, int containerViewId, int pos, FragmentTransaction transaction, boolean isRoot, SupportFragment[] fragments) {
        if (fragments != null && fragments.length > 0) {
            if (pos >= fragments.length || pos < 0) {
                throw new IndexOutOfBoundsException("Index: " + pos + ", Size: " + fragments.length);
            }
            for (int i = 0; i < fragments.length; i++) {
                SupportFragment f = fragments[i];
                f.isRoot = isRoot;
                bindContainerId(containerViewId, f);
                String tag = f.getClass().getSimpleName();
                transaction.add(containerViewId, f, tag)
                        .addToBackStack(tag);

                if (i == pos) {
                    transaction.show(f);
                } else {
                    transaction.hide(f);
                }
            }
            transaction.commit();
        }
    }

    private void bindContainerId(int containerId, SupportFragment to) {
        Bundle args = to.getArguments();
        if (args == null) {
            args = new Bundle();
            to.setArguments(args);
        }
        args.putInt(SupportFragment.ARG_CONTAINER, containerId);
    }

    public void initFragments(Bundle savedInstanceState, SupportFragment fragment) {
        if (savedInstanceState == null) {
            return;
        }
        boolean isSupportHidden = savedInstanceState.getBoolean(SupportFragment.ARG_IS_HIDDEN);

        FragmentTransaction ft = manager.beginTransaction();
        if (isSupportHidden) {
            ft.hide(fragment);
        } else {
            ft.show(fragment);
        }
        ft.commit();
    }


    public boolean popTo(Class<? extends SupportFragment> target, boolean includeSelf) {
        int flag;
        if (includeSelf) {
            flag = FragmentManager.POP_BACK_STACK_INCLUSIVE;
        } else {
            flag = 0;
        }
        return manager.popBackStackImmediate(target.getSimpleName(), flag);
    }

    public boolean popSelf() {
        return manager.popBackStackImmediate();
    }

    public <T extends SupportFragment> T findFragment(@NonNull Class<T> className) {
        Fragment tagFragment = manager.findFragmentByTag(className.getSimpleName());
        try {
            return (T) tagFragment;
        } catch (Exception e) {
            return null;
        }
    }

    @Nullable
    public List<Fragment> getTopFragments() {
        List<Fragment> fragments = manager.getFragments();
        List<Fragment> topFragments = new ArrayList<>();
        if (fragments == null) {
            return null;
        }
        int size = fragments.size();
        for (int i = size - 1; i >= 0; i--) {
            Fragment f = fragments.get(i);
            if (!f.isHidden()) {
                Fragment t = getTopFragment(f.getChildFragmentManager());//递归
                if (t != null) {
                    topFragments.add(t);
                } else {
                    topFragments.add(f);
                }
            }

        }
        return topFragments;
    }

    @Nullable
    private Fragment getTopFragment(FragmentManager manager) {
        List<Fragment> fragments = manager.getFragments();
        if (fragments == null) {
            return null;
        }
        int size = fragments.size();
        for (int i = size - 1; i >= 0; i--) {
            Fragment f = fragments.get(i);
            if (!f.isHidden()) {
                Fragment tTopFragment = getTopFragment(f.getChildFragmentManager());
                return tTopFragment == null ? f : tTopFragment;
            }
        }
        return null;
    }
}

File: app/src/main/java/com/lovejjfg/arsenal/utils/FragmentsUtil.kt
/*
 *  Copyright (c) 2017.  Joe
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

package com.lovejjfg.arsenal.utils

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentTransaction

import com.lovejjfg.arsenal.base.SupportFragment

import java.util.ArrayList

/**
 * Created by Joe on 2016/11/13.
 * Email lovejjfg@gmail.com
 */

class FragmentsUtil(private val manager: FragmentManager?) {

    //递归
    val topFragments: List<Fragment>?
        get() {
            if (manager == null) {
                return null
            }
            val fragments = manager.fragments
            val topFragments = ArrayList<Fragment>()
            if (fragments.isEmpty()) {
                return null
            }
            val size = fragments.size
            for (i in size - 1 downTo 0) {
                val f = fragments[i]
                if (f != null && f.isAdded && !f.isHidden && f.userVisibleHint) {
                    val t = getTopFragment(f.childFragmentManager)
                    if (t != null) {
                        topFragments.add(t)
                    } else {
                        topFragments.add(f)
                    }
                }
            }
            return topFragments
        }

    fun addToShow(from: SupportFragment, to: SupportFragment) {
        bindContainerId(from.containerId, to)
        val transaction = manager!!.beginTransaction()
        val tag = to.javaClass.simpleName
        transaction.add(from.containerId, to, tag)
            .addToBackStack(tag)
            .hide(from)
            .show(to)
            .commit()
    }

    fun addToParent(containerViewId: Int, parent: SupportFragment, pos: Int, vararg children: SupportFragment) {
        val transaction = parent.childFragmentManager.beginTransaction()
        if (children.isNotEmpty()) {
            addFragmentsToStack(containerViewId, pos, transaction, false, children)
        }
    }

    fun replaceToParent(containerViewId: Int, parent: SupportFragment, vararg childs: SupportFragment) {
        val transaction = parent.childFragmentManager.beginTransaction()
        if (childs.isNotEmpty()) {
            for (child in childs) {
                bindContainerId(containerViewId, child)
                val tag = child.javaClass.simpleName
                transaction.replace(containerViewId, child, tag)
                    .addToBackStack(tag)
            }
            transaction.commit()
        }
    }

    fun replaceToShow(from: SupportFragment, to: SupportFragment) {
        bindContainerId(from.containerId, to)
        val transaction = manager!!.beginTransaction()
        val tag = to.javaClass.simpleName
        transaction.replace(from.containerId, to, tag)
            .addToBackStack(tag)
            .commit()
    }

    fun loadRoot(containerViewId: Int, pos: Int, vararg roots: SupportFragment) {
        val transaction = manager!!.beginTransaction()
        addFragmentsToStack(containerViewId, pos, transaction, true, roots)
    }

    private fun addFragmentsToStack(
        containerViewId: Int,
        pos: Int,
        transaction: FragmentTransaction,
        isRoot: Boolean,
        fragments: Array<out SupportFragment>
    ) {
        if (fragments.isNotEmpty()) {
            if (pos >= fragments.size || pos < 0) {
                throw IndexOutOfBoundsException("Index: " + pos + ", Size: " + fragments.size)
            }
            for (i in fragments.indices) {
                val f = fragments[i]
                f.isRoot = isRoot
                bindContainerId(containerViewId, f)
                val tag = f.javaClass.simpleName
                transaction.add(containerViewId, f, tag)
                    .addToBackStack(tag)

                if (i == pos) {
                    transaction.show(f)
                } else {
                    transaction.hide(f)
                }
            }
            transaction.commit()
        }
    }

    private fun bindContainerId(containerId: Int, to: SupportFragment) {
        var args = to.arguments
        if (args == null) {
            args = Bundle()
            to.arguments = args
        }
        args.putInt(SupportFragment.ARG_CONTAINER, containerId)
    }

    fun initFragments(savedInstanceState: Bundle?, fragment: SupportFragment) {
        if (savedInstanceState == null) {
            return
        }
        val isSupportHidden = savedInstanceState.getBoolean(SupportFragment.ARG_IS_HIDDEN)

        val ft = manager!!.beginTransaction()
        if (isSupportHidden) {
            ft.hide(fragment)
        } else {
            ft.show(fragment)
        }
        ft.commit()
    }

    fun popTo(target: Class<out SupportFragment>, includeSelf: Boolean): Boolean {
        val flag: Int = if (includeSelf) {
            FragmentManager.POP_BACK_STACK_INCLUSIVE
        } else {
            0
        }
        return manager!!.popBackStackImmediate(target.simpleName, flag)
    }

    fun popSelf(): Boolean {
        return manager!!.popBackStackImmediate()
    }

    fun <T : SupportFragment> findFragment(className: Class<T>): T? {
        return findFragment(null, className)
    }

    fun <T : SupportFragment> findFragment(parentFragment: SupportFragment?, className: Class<T>): T? {
        val tagFragment: Fragment? = if (parentFragment != null) {
            parentFragment.childFragmentManager.findFragmentByTag(className.simpleName)
        } else {
            manager!!.findFragmentByTag(className.simpleName)
        }
        return try {
            tagFragment as T?
        } catch (e: Exception) {
            null
        }
    }

    private fun getTopFragment(manager: FragmentManager): Fragment? {
        val fragments = manager.fragments
        if (fragments.isEmpty()) {
            return null
        }
        val size = fragments.size
        for (i in size - 1 downTo 0) {
            val f = fragments[i]
            if (f != null && f.isAdded && !f.isHidden && f.userVisibleHint) {
                val tTopFragment = getTopFragment(f.childFragmentManager)
                return tTopFragment ?: f
            }
        }
        return null
    }
}
